File: compilers/python/tests/test_ir_loader_size_guard.py
"""BUG-008 follow-up: IR-loader size-guard regression tests."""

import pytest

from runar_compiler.ir.input_limits import (
    MAX_IR_BYTES,
    MAX_IR_NESTING,
    IRNestingExceededError,
    IRSizeExceededError,
    assert_ir_bytes_under_limit,
    assert_ir_nesting_under_limit,
)
from runar_compiler.ir.loader import load_ir


def test_load_ir_rejects_oversized_input() -> None:
    oversized = " " * (MAX_IR_BYTES + 1)
    with pytest.raises(IRSizeExceededError) as exc:
        load_ir(oversized)
    assert exc.value.limit == MAX_IR_BYTES
    assert exc.value.actual == MAX_IR_BYTES + 1


def test_load_ir_rejects_deeply_nested_input() -> None:
    depth = MAX_IR_NESTING + 50
    body = "1"
    for _ in range(depth):
        body = '{"n":' + body + "}"
    with pytest.raises(IRNestingExceededError) as exc:
        load_ir(body)
    assert exc.value.limit == MAX_IR_NESTING


def test_depth_walk_ignores_braces_inside_strings() -> None:
    # 1000 `{` inside a JSON string MUST NOT count toward depth.
    inner = "{" * 1000
    bad = (
        '{"contractName":"X","properties":[],"methods":[],'
        f'"_note":"{inner}"}}'
    )
    # Neither DoS-bound guard should fire. The downstream loader might
    # raise ValueError on missing fields or otherwise succeed — neither
    # IRSizeExceededError nor IRNestingExceededError should appear.
    try:
        load_ir(bad)
    except (IRSizeExceededError, IRNestingExceededError) as e:
        pytest.fail(f"DoS-bound guard incorrectly tripped: {e}")
    except ValueError:
        pass  # downstream schema/decoding may fail; that's fine.


def test_load_ir_accepts_minimal_program() -> None:
    minimal = '{"contractName":"X","properties":[],"methods":[]}'
    program = load_ir(minimal)
    assert program is not None


def test_assert_ir_bytes_under_limit_typed_error() -> None:
    with pytest.raises(IRSizeExceededError):
        assert_ir_bytes_under_limit("x" * (MAX_IR_BYTES + 1))


def test_assert_ir_nesting_under_limit_typed_error() -> None:
    depth = MAX_IR_NESTING + 1
    body = "1"
    for _ in range(depth):
        body = "[" + body + "]"
    with pytest.raises(IRNestingExceededError):
        assert_ir_nesting_under_limit(body)

File: compilers/ruby/test/test_ir_loader_size_guard.rb
# frozen_string_literal: true

# BUG-008 follow-up: IR-loader size-guard regression tests.

require_relative "test_helper"
require "runar_compiler/ir/input_limits"
require "runar_compiler/ir/loader"

class TestIRLoaderSizeGuard < Minitest::Test
  MAX_BYTES = RunarCompiler::IR::InputLimits::MAX_IR_BYTES
  MAX_DEPTH = RunarCompiler::IR::InputLimits::MAX_IR_NESTING
  SIZE_ERR = RunarCompiler::IR::InputLimits::IRSizeExceededError
  NEST_ERR = RunarCompiler::IR::InputLimits::IRNestingExceededError

  def test_load_ir_rejects_oversized_input
    oversized = " " * (MAX_BYTES + 1)
    err = assert_raises(SIZE_ERR) do
      RunarCompiler::IR.load_ir(oversized)
    end
    assert_equal MAX_BYTES, err.limit
    assert_equal MAX_BYTES + 1, err.actual
  end

  def test_load_ir_rejects_deeply_nested_input
    depth = MAX_DEPTH + 50
    body = "1"
    depth.times { body = '{"n":' + body + "}" }
    err = assert_raises(NEST_ERR) do
      RunarCompiler::IR.load_ir(body)
    end
    assert_equal MAX_DEPTH, err.limit
  end

  def test_depth_walk_ignores_braces_inside_strings
    open_braces = "{" * 1000
    bad =
      '{"contractName":"X","properties":[],"methods":[],"_note":"' +
      open_braces + '"}'
    # Neither DoS-bound guard should fire. Downstream parse should
    # succeed (X has the required shape).
    program = RunarCompiler::IR.load_ir(bad)
    assert_equal "X", program.contract_name
  end

  def test_load_ir_accepts_minimal_program
    minimal = '{"contractName":"X","properties":[],"methods":[]}'
    program = RunarCompiler::IR.load_ir(minimal)
    assert_equal "X", program.contract_name
  end

  def test_assert_ir_bytes_under_limit_typed_error
    assert_raises(SIZE_ERR) do
      RunarCompiler::IR::InputLimits.assert_ir_bytes_under_limit("x" * (MAX_BYTES + 1))
    end
  end

  def test_assert_ir_nesting_under_limit_typed_error
    body = "1"
    (MAX_DEPTH + 1).times { body = "[" + body + "]" }
    assert_raises(NEST_ERR) do
      RunarCompiler::IR::InputLimits.assert_ir_nesting_under_limit(body)
    end
  end
end
